### PR TITLE
Add an audit database logging resource changes

### PR DESCRIPTION
**Pull request by pallix**

### CODE REVIEW

**Comment by joka** [comment]
Its nicer to review if you separate the commits, like:

- use request finish callback to clear the transaction log
- add auditlog to ...
- regenerate sphinx api doc

To do this you can use interactive staging or rebasing

File: etc/noserver.ini.in [hunk 1]
```diff
@@ -23,6 +23,7 @@ pyramid.includes =
 tm.attempts = 5
 
 zodbconn.uri = memory://
+zodbconn.uri.audit = memory://auditlog
```
Comment (joka) [inline etc/noserver.ini.in]: nosever is only used by buildout to autogenerate typescript classed, 
no need for auditlog

But we should add all new config parameters in the example development init in  adhocracore_core.scaffold.--

File: etc/test.ini.in [hunk 2]
```diff
@@ -10,6 +10,7 @@ pyramid.includes =
     pyramid_tm
     pyramid_mailer
 zodbconn.uri = memory://
+zodbconn.uri.audit = memory://auditlog
 adhocracy.add_test_users = true
 substanced.secret = seekri1
```
Comment (joka) [inline etc/test.ini.in]: same here, audit log does not hurt but is not needed

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+"""Auditlog of events stored in a ZODB database."""
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: may more informative:

    Log which user modifies resources in additional 'audit' database.

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+from adhocracy_core.utils import get_sheet_field
+
+
+logger = getLogger(__name__)
+
+AuditEntry = namedtuple('AuditEntry', ['name', 'payload'])
+
+RESOURCE_CREATED = 'resourceCreated'
+RESOURCE_MODIFIED = 'resourceModified'
+
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: the resource suffice is not needed,  we know it about resources

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    """An Auditlog composed of entries."""
+
+    def add(self, name, **kw):
+        """ Add a record the audit log.
+
+        ``_name`` should be the event name,
+        ``_oid`` should be an object oid or ``None``, and ``kw`` should be a
+        json-serializable dictionary.
+        """
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: methods shoud have type annotations (similar syntax like  mypy) for documentation and static type checking.  
If the method arguments are not self explaining or the functions raises Exception you can add additional information with javadoc-style in the docstring.

Why does this function accept arbitrary keywords?

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        payload = json.dumps(kw)
+        self[datetime.utcnow()] = AuditEntry(name, payload)
+
+
+def get_auditlog(context):
+    """Return the auditlog. Create one if none exits."""
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: I know i came up with this, but in genrell a low  function should do one "thing" like create, edit, delete,...

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def log_auditevent(context, name, **kw):
+    """Add a an audit entry to the audit database.
+
+    The audit database is created if missing. If the zodbconn.uri.audit
+    value is not specified in the config, auditing does not happen.
+    """
+    auditlog = get_auditlog(context)
+    if auditlog is not None:
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: To prevent indention we do in most cases:
 
    if x is None:
        return
    continue with programm flow

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                       user_name=user_name,
+                       user_path=user_path)
+    elif change.modified:
+        log_auditevent(context,
+                       RESOURCE_MODIFIED,
+                       resource_path=path,
+                       user_name=user_name,
+                       user_path=user_path)
+    # else: log visibility changes?
+    transaction.commit()
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: is this commit neccessary?

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                       resource_path=path,
+                       user_name=user_name,
+                       user_path=user_path)
+    elif change.modified:
+        log_auditevent(context,
+                       RESOURCE_MODIFIED,
+                       resource_path=path,
+                       user_name=user_name,
+                       user_path=user_path)
+    # else: log visibility changes?
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: yip, we should track visibility changes

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def audit_changes_callback(request, response):
+    """Add audit entries to the auditlog when the data is changed."""
+    registry = request.registry
+    changelog_metadata = registry.changelog.values()
+    if len(changelog_metadata) == 0:
+        return
+
+    (user_name, user_path) = _get_user_info(request)
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: the ( , and ) is not neede

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    if auditlog is not None:
+        auditlog.add(name, **kw)
+
+
+def audit_changes_callback(request, response):
+    """Add audit entries to the auditlog when the data is changed."""
+    registry = request.registry
+    changelog_metadata = registry.changelog.values()
+    if len(changelog_metadata) == 0:
+        return
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: this statement is not need, the loop below can handle a empty list

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py (hunk 10 above)
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: Why did you add the empty lines?

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    The audit database is created if missing. If the zodbconn.uri.audit
+    value is not specified in the config, auditing does not happen.
+    """
+    auditlog = get_auditlog(context)
+    if auditlog is not None:
+        auditlog.add(name, **kw)
+
+
+def audit_changes_callback(request, response):
+    """Add audit entries to the auditlog when the data is changed."""
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: I did not understand at once what this function does, so I would add the hind "finished_callback" to the function name or the docstring.

File: src/adhocracy_core/adhocracy_core/auditing/test_init.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+
+@fixture()
+def request_(registry):
+    request = testing.DummyResource(registry=registry)
+    return request
+
+
+def test_add_events(context):
+    from . import log_auditevent
```
Comment (joka) [inline src/adhocracy_core/adhocracy_core/auditing/test_init.py]: Its a bit unclear what do want to test here and why do you slow down the test with 10000 entries?

A unit test should be named after the test condition or  the expected behavior.
The program code should follow the structure "given" "when", "then"
and test one "thing" (preferred one assert statement)

Also in the following test its not always clear what function is under test (get_auditlog or log_auditevent).

**Comment by joka** [comment]
I added a lot of style guide comments. This must not be all done before merge, but should be done when we do refactoring.

**Comment by joka** [comment]
the auditing package has 67% test coverage, this should  be 100%
we should do more test driven development ;)

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py (hunk 4 above)
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: you mean only in the strings not in the variables names right?

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py (hunk 12 above)
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: should I renamed the function into audit_resource_changes_callback?

File: src/adhocracy_core/adhocracy_core/auditing/__init__.py (hunk 8 above)
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/auditing/__init__.py]: yes, the commit() is necessaru

**Comment by pallix** [comment]
I'm rather sure I always had 100% coverage of my added code :-)

**Comment by pallix** [comment]
I'm done with my changes.

**Comment by pallix** [comment]
Actually I forgot two changes

**Comment by joka** [comment]
can you merge master?

**Comment by pallix** [comment]
merge from master is done

**Comment by pallix** [comment]
LGTM!

**Comment by joka** [comment]
juhu
